opentelemetry-sdk/tests/metrics/test_metrics.py

```diff
@@ -27,35 +27,38 @@ def test_extends_api(self):
     def test_record_batch(self):
         meter = metrics.Meter()
         label_keys = ("key1",)
-        label_values = ("value1",)
         counter = metrics.Counter("name", "desc", "unit", float, label_keys)
+        kvp = {"key1": "value1"}
+        label_set = meter.get_label_set(kvp)
         record_tuples = [(counter, 1.0)]
-        meter.record_batch(label_values, record_tuples)
-        self.assertEqual(counter.get_handle(label_values).data, 1.0)
+        meter.record_batch(label_set, record_tuples)
+        self.assertEqual(counter.get_handle(label_set).data, 1.0)
 
     def test_record_batch_multiple(self):
         meter = metrics.Meter()
         label_keys = ("key1", "key2", "key3")
-        label_values = ("value1", "value2", "value3")
+        kvp = {"key1": "value1", "key2": "value2", "key3": "value3"}
+        label_set = meter.get_label_set(kvp)
         counter = metrics.Counter("name", "desc", "unit", float, label_keys)
         gauge = metrics.Gauge("name", "desc", "unit", int, label_keys)
         measure = metrics.Measure("name", "desc", "unit", float, label_keys)
         record_tuples = [(counter, 1.0), (gauge, 5), (measure, 3.0)]
-        meter.record_batch(label_values, record_tuples)
-        self.assertEqual(counter.get_handle(label_values).data, 1.0)
-        self.assertEqual(gauge.get_handle(label_values).data, 5)
-        self.assertEqual(measure.get_handle(label_values).data, 0)
+        meter.record_batch(label_set, record_tuples)
+        self.assertEqual(counter.get_handle(label_set).data, 1.0)
+        self.assertEqual(gauge.get_handle(label_set).data, 5)
+        self.assertEqual(measure.get_handle(label_set).data, 0)
 
     def test_record_batch_exists(self):
         meter = metrics.Meter()
         label_keys = ("key1",)
-        label_values = ("value1",)
+        kvp = {"key1": "value1"}
+        label_set = meter.get_label_set(kvp)
         counter = metrics.Counter("name", "desc", "unit", float, label_keys)
-        counter.add(label_values, 1.0)
-        handle = counter.get_handle(label_values)
+        counter.add(label_set, 1.0)
```

Comment:
will these methods also support standard dicts as well as LabelSets? I think we should.

Looking at https://github.com/open-telemetry/oteps/pull/49/files it looks like the primary motivation for LabelSet is performance. I don't see a lot here that will enable better performance, and it's adding another layer of abstraction which introduces complexity.

Reply:
The primary motivation for LabelSet is performance, yes. It is the canonicalization of the label keys/values into unique identifiers that is very expensive. This is why LabelSets are created once using specific label keys/values, which are then canonicalized once, and then are able to be used multiple times without having to perform this expensive action. In the implementation, this is demonstrated by using a dictionary of encoded strings, to find the corresponding metric handle that was created from the canonicalized label keys/values.

Your suggestion of supporting standard dicts is valid though, the underlying logic will just take in the dict and construct a Labelset. This is more for convenience rather than an performance improvement. However, the [Go](https://github.com/open-telemetry/opentelemetry-go/blob/master/sdk/metric/sdk.go#L161) and [JavaScript](https://github.com/open-telemetry/opentelemetry-js/blob/3b240cb0e824526cd399d3da91bbecaa83af25a9/packages/opentelemetry-metrics/src/Metric.ts#L54) implementations only allow passing in LabelSets for now so I'd like to stay consistent. Good idea for the future though.

Reply:
Not to get into the project approach, but I think half the purpose of implementing in multiple languages is to identify patterns that work or don't for various languages. I think we should feel empowered to contribute feedback around the approach upstream.

That all said, I understand the underlying philosophy that using an existing labelset will omit some processing that needs to happen every time. If I understand correctly, you're saying we save this by pre-encoding the strings.

I'm not really sure that pre-encoding the strings saves us any time here. Maybe feedback for the statsite encoding side, but that encoding makes the assumption that whatever is consuming the key-value pairs understands the statsite encoding and can optimize on that, which I don't believe is true.

For example, if we implement some sort of processor that needs to parse and filter the key-value pairs, then work will have to occur to parse out the key-value pairs, and reconstruct them into a dictionary or some secondary data structure to parse out the values. That negates the performance benefit when compared to just using a dict.

Reply:
I believe this was addressed with the exporters utilizing specific encodings for optimization. With introducing of an "encoding layer", where users can configure their own encodings I believe this would save some time.

Reply:
@toumorokoshi: out of curiosity, how would you implement this? Make `label_set` a `Union[Dict[str, str], LabelSet]` or have separate methods for each type?

@lzchen and I talked about it, and this seems to be a case where the conventions of the language shape the API. Java makes it easy to support both arg types with overloading, python doesn't.

That said, I'd personally prefer that we didn't *require* `LabelSets` anywhere in the API. This looks like a bit of premature optimization on the part of the spec, at the expense of the simplicity of the API. Especially when we consider applications that don't reuse labels. But this is a problem for the spec, not this PR.